# Section : 52S

# Course name and number : 
Information Retrieval (DS321)

# Project title :
What is Data science ?

# Group members :

Sarah Fahad Alotaibi                        
443007420

Yasmeen Ammar Almutairi                    
443007424

# Brief description and goal of analysis:
The goal of building an information retrieval (IR) project is to develop system and algorithms that efficiently and effectively retrieve relevant information from a large collection of data based on user queries or information needs. The primary objectives of our information retrieval project, which the query here is "what is data science "and the system we built should retrive the relevent document among the 3 documents we provide to the system .

# The proposed corpus and query

In [68]:
Doc1 = """ Data science combines math and statistics , specialized programming , advanced analytics , These insights can be used to guide decision making and strategic planning """

In [69]:
Doc2 = """ Machine learning with specific subject matter expertise to uncover actionable , insights hidden in an organization data """

In [70]:
Doc3= """ The accelerating volume of data sources, and subsequently data , has made data science is one of the fastest growing field across every industry """

In [71]:
Q ="""What is Data science?"""

# Data Preprocessing 
The preprocessing process: its preparation and preprocessing  methods, this procedure starts with taking an input of raw text and return cleansed tokens.

# converting to lowercase : 
Change the upper case words to lowercase

In [72]:
#converting to lowercase using a function 
Doc1 = Doc1.lower()
Doc1

' data science combines math and statistics , specialized programming , advanced analytics , these insights can be used to guide decision making and strategic planning '

In [73]:
Doc2 = Doc2.lower()
Doc2

' machine learning with specific subject matter expertise to uncover actionable , insights hidden in an organization data '

In [74]:
Doc3 = Doc3.lower()
Doc3

' the accelerating volume of data sources, and subsequently data , has made data science is one of the fastest growing field across every industry '

# Remove punctuation

In [75]:
#Using  the library String to use  a method in this library to remove punctuation
import string

In [76]:
# Remove the punctuatuon from documents 
Doc1 = Doc1.translate(str.maketrans("", "",string.punctuation))
Doc1

' data science combines math and statistics  specialized programming  advanced analytics  these insights can be used to guide decision making and strategic planning '

In [77]:
Doc2 = Doc2.translate(str.maketrans("", "",string.punctuation))
Doc2

' machine learning with specific subject matter expertise to uncover actionable  insights hidden in an organization data '

In [78]:
Doc3 = Doc3.translate(str.maketrans("", "",string.punctuation))
Doc3

' the accelerating volume of data sources and subsequently data  has made data science is one of the fastest growing field across every industry '

# Tokenizing 

Tokenizing is the process of breaking down a text into individual units, called tokens. In the context of natural language processing (NLP)

In [79]:
# use the library nltk which is The Natural Language Toolkit apply preprocessing 
#(NLTK provides a wide range of text processing utilities, including tokenization, stemming, lemmatization)
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\Asus\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [80]:
from nltk.tokenize import word_tokenize #Tokenize

In [81]:
# turning the documents into tokens 
tokens1= nltk.word_tokenize(Doc1)
tokens1

['data',
 'science',
 'combines',
 'math',
 'and',
 'statistics',
 'specialized',
 'programming',
 'advanced',
 'analytics',
 'these',
 'insights',
 'can',
 'be',
 'used',
 'to',
 'guide',
 'decision',
 'making',
 'and',
 'strategic',
 'planning']

In [82]:
tokens2= nltk.word_tokenize(Doc2)
tokens2

['machine',
 'learning',
 'with',
 'specific',
 'subject',
 'matter',
 'expertise',
 'to',
 'uncover',
 'actionable',
 'insights',
 'hidden',
 'in',
 'an',
 'organization',
 'data']

In [83]:
tokens3= nltk.word_tokenize(Doc3)
tokens3

['the',
 'accelerating',
 'volume',
 'of',
 'data',
 'sources',
 'and',
 'subsequently',
 'data',
 'has',
 'made',
 'data',
 'science',
 'is',
 'one',
 'of',
 'the',
 'fastest',
 'growing',
 'field',
 'across',
 'every',
 'industry']

# Removing stop words
Stop Words pre-processing:  to drop frequent used words, which add no value to the analysis.

In [84]:
# Useing stop words so we can Improve Search Accuracy
from nltk.corpus import stopwords #stop words

In [85]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Asus\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [86]:
english_stop=set(stopwords.words('english'))
english_stop

{'a',
 'about',
 'above',
 'after',
 'again',
 'against',
 'ain',
 'all',
 'am',
 'an',
 'and',
 'any',
 'are',
 'aren',
 "aren't",
 'as',
 'at',
 'be',
 'because',
 'been',
 'before',
 'being',
 'below',
 'between',
 'both',
 'but',
 'by',
 'can',
 'couldn',
 "couldn't",
 'd',
 'did',
 'didn',
 "didn't",
 'do',
 'does',
 'doesn',
 "doesn't",
 'doing',
 'don',
 "don't",
 'down',
 'during',
 'each',
 'few',
 'for',
 'from',
 'further',
 'had',
 'hadn',
 "hadn't",
 'has',
 'hasn',
 "hasn't",
 'have',
 'haven',
 "haven't",
 'having',
 'he',
 'her',
 'here',
 'hers',
 'herself',
 'him',
 'himself',
 'his',
 'how',
 'i',
 'if',
 'in',
 'into',
 'is',
 'isn',
 "isn't",
 'it',
 "it's",
 'its',
 'itself',
 'just',
 'll',
 'm',
 'ma',
 'me',
 'mightn',
 "mightn't",
 'more',
 'most',
 'mustn',
 "mustn't",
 'my',
 'myself',
 'needn',
 "needn't",
 'no',
 'nor',
 'not',
 'now',
 'o',
 'of',
 'off',
 'on',
 'once',
 'only',
 'or',
 'other',
 'our',
 'ours',
 'ourselves',
 'out',
 'over',
 'own',
 'r

In [87]:
# removing stop words from document one 
filtered_words1=[word for word in tokens1 if word not in stopwords.words('english')]
filtered_words1_set=set(filtered_words1)
print(filtered_words1_set)

{'math', 'guide', 'strategic', 'programming', 'science', 'insights', 'used', 'planning', 'data', 'making', 'combines', 'statistics', 'decision', 'advanced', 'analytics', 'specialized'}


In [88]:
# removing stop words from document two
filtered_words2=[word for word in tokens2 if word not in stopwords.words('english')]
filtered_words2_set=set(filtered_words2)
print(filtered_words2_set)

{'subject', 'organization', 'insights', 'specific', 'expertise', 'matter', 'actionable', 'data', 'hidden', 'uncover', 'learning', 'machine'}


In [89]:
# removing stop words from document three
filtered_words3=[word for word in tokens3 if word not in stopwords.words('english')]
filtered_words3_set=set(filtered_words3)
print(filtered_words3_set)

{'accelerating', 'science', 'one', 'subsequently', 'growing', 'sources', 'field', 'data', 'made', 'every', 'fastest', 'industry', 'across', 'volume'}


# Stemming
Stemming is reducing words to their base or root form so we can improve the search results.

In [91]:
from nltk.stem import PorterStemmer #Stemming

In [92]:
word_stemmer = PorterStemmer() # this method perform stemming 

In [93]:
# apply stemming on the documents 
stemmed_words1 = [word_stemmer.stem(word) for word in filtered_words1]
print(stemmed_words1)

['data', 'scienc', 'combin', 'math', 'statist', 'special', 'program', 'advanc', 'analyt', 'insight', 'use', 'guid', 'decis', 'make', 'strateg', 'plan']


In [94]:
stemmed_words2 = [word_stemmer.stem(word) for word in filtered_words2]
print(stemmed_words2)

['machin', 'learn', 'specif', 'subject', 'matter', 'expertis', 'uncov', 'action', 'insight', 'hidden', 'organ', 'data']


In [95]:
stemmed_words3 = [word_stemmer.stem(word) for word in filtered_words3]
print(stemmed_words3)

['acceler', 'volum', 'data', 'sourc', 'subsequ', 'data', 'made', 'data', 'scienc', 'one', 'fastest', 'grow', 'field', 'across', 'everi', 'industri']


In [96]:
#spliting the query to improve the results 
Q_words = Q.split(' ')
Q_words

['What', 'is', 'Data', 'science?']

In [97]:
#apply stemming on the query for calculating the TF-IDF 
stemmed_wordsQ = [word_stemmer.stem(word) for word in Q_words]
print(stemmed_wordsQ)

['what', 'is', 'data', 'science?']


# Term Frequency (TF)
Term frequency is the measurement of how frequently a term occurs within a document.

In [98]:
#compining all the documents so we can calculate the term frequency and accurate values 
uniqueWords = set(stemmed_words1).union(set(stemmed_words2)).union(set(stemmed_words3))
uniqueWordQ = set(stemmed_wordsQ) #  calculate the term frequency for the query 

In [99]:
print(uniqueWords)
print(f'Unique Word Question {uniqueWordQ}')

{'subject', 'plan', 'fastest', 'across', 'acceler', 'machin', 'math', 'action', 'statist', 'learn', 'one', 'specif', 'grow', 'advanc', 'data', 'guid', 'subsequ', 'uncov', 'combin', 'special', 'matter', 'make', 'organ', 'field', 'made', 'everi', 'hidden', 'industri', 'volum', 'sourc', 'expertis', 'use', 'insight', 'program', 'analyt', 'strateg', 'decis', 'scienc'}
Unique Word Question {'data', 'is', 'science?', 'what'}


In [100]:
num_of_word1=dict.fromkeys(uniqueWords,0)
for word in stemmed_words1:
  num_of_word1[word]+=1
num_of_word2=dict.fromkeys(uniqueWords,0)
for word in stemmed_words2:
  num_of_word2[word]+=1
num_of_word3=dict.fromkeys(uniqueWords,0)
for word in stemmed_words3:
  num_of_word3[word]+=1
num_of_wordQ=dict.fromkeys(uniqueWordQ,0)
for word in stemmed_wordsQ:
  num_of_wordQ[word]+=1

In [101]:
#TF of document one 
print(num_of_word1)

{'subject': 0, 'plan': 1, 'fastest': 0, 'across': 0, 'acceler': 0, 'machin': 0, 'math': 1, 'action': 0, 'statist': 1, 'learn': 0, 'one': 0, 'specif': 0, 'grow': 0, 'advanc': 1, 'data': 1, 'guid': 1, 'subsequ': 0, 'uncov': 0, 'combin': 1, 'special': 1, 'matter': 0, 'make': 1, 'organ': 0, 'field': 0, 'made': 0, 'everi': 0, 'hidden': 0, 'industri': 0, 'volum': 0, 'sourc': 0, 'expertis': 0, 'use': 1, 'insight': 1, 'program': 1, 'analyt': 1, 'strateg': 1, 'decis': 1, 'scienc': 1}


In [102]:
#TF of document two
print(num_of_word2)

{'subject': 1, 'plan': 0, 'fastest': 0, 'across': 0, 'acceler': 0, 'machin': 1, 'math': 0, 'action': 1, 'statist': 0, 'learn': 1, 'one': 0, 'specif': 1, 'grow': 0, 'advanc': 0, 'data': 1, 'guid': 0, 'subsequ': 0, 'uncov': 1, 'combin': 0, 'special': 0, 'matter': 1, 'make': 0, 'organ': 1, 'field': 0, 'made': 0, 'everi': 0, 'hidden': 1, 'industri': 0, 'volum': 0, 'sourc': 0, 'expertis': 1, 'use': 0, 'insight': 1, 'program': 0, 'analyt': 0, 'strateg': 0, 'decis': 0, 'scienc': 0}


In [103]:
#TF of document three
print(num_of_word3)

{'subject': 0, 'plan': 0, 'fastest': 1, 'across': 1, 'acceler': 1, 'machin': 0, 'math': 0, 'action': 0, 'statist': 0, 'learn': 0, 'one': 1, 'specif': 0, 'grow': 1, 'advanc': 0, 'data': 3, 'guid': 0, 'subsequ': 1, 'uncov': 0, 'combin': 0, 'special': 0, 'matter': 0, 'make': 0, 'organ': 0, 'field': 1, 'made': 1, 'everi': 1, 'hidden': 0, 'industri': 1, 'volum': 1, 'sourc': 1, 'expertis': 0, 'use': 0, 'insight': 0, 'program': 0, 'analyt': 0, 'strateg': 0, 'decis': 0, 'scienc': 1}


In [104]:
#TF of the query 
print(num_of_wordQ)

{'data': 1, 'is': 1, 'science?': 1, 'what': 1}


In [105]:
# to format the results and make them look readable 
import pandas as pd
pd.DataFrame([num_of_word1,num_of_word2,num_of_word3])

subject  plan  fastest  across  acceler  machin  math  action  statist  \
0        0     1        0       0        0       0     1       0        1   
1        1     0        0       0        0       1     0       1        0   
2        0     0        1       1        1       0     0       0        0   

   learn  ...  volum  sourc  expertis  use  insight  program  analyt  strateg  \
0      0  ...      0      0         0    1        1        1       1        1   
1      1  ...      0      0         1    0        1        0       0        0   
2      0  ...      1      1         0    0        0        0       0        0   

   decis  scienc  
0      1       1  
1      0       0  
2      0       1  

[3 rows x 38 columns]

In [106]:
pd.DataFrame([num_of_wordQ])

data  is  science?  what
0     1   1         1     1

In [107]:
#TF, this method calculate the term frequency 
def computeTF(word_dict,list_of_word):
  TF_Dict ={}
  word_count = len(list_of_word)
  # you just pass the word with [0] => And you must use [word]
  for word,count in word_dict.items():
    TF_Dict[word] = count / float(word_count)
  return TF_Dict

In [108]:
# using the method that calculate the term frequency for document one 
TF_Doc1 = computeTF(num_of_word1,stemmed_words1)
TF_Doc1

{'subject': 0.0,
 'plan': 0.0625,
 'fastest': 0.0,
 'across': 0.0,
 'acceler': 0.0,
 'machin': 0.0,
 'math': 0.0625,
 'action': 0.0,
 'statist': 0.0625,
 'learn': 0.0,
 'one': 0.0,
 'specif': 0.0,
 'grow': 0.0,
 'advanc': 0.0625,
 'data': 0.0625,
 'guid': 0.0625,
 'subsequ': 0.0,
 'uncov': 0.0,
 'combin': 0.0625,
 'special': 0.0625,
 'matter': 0.0,
 'make': 0.0625,
 'organ': 0.0,
 'field': 0.0,
 'made': 0.0,
 'everi': 0.0,
 'hidden': 0.0,
 'industri': 0.0,
 'volum': 0.0,
 'sourc': 0.0,
 'expertis': 0.0,
 'use': 0.0625,
 'insight': 0.0625,
 'program': 0.0625,
 'analyt': 0.0625,
 'strateg': 0.0625,
 'decis': 0.0625,
 'scienc': 0.0625}

In [109]:
# using the method that calculate the term frequency for document two
TF_Doc2 = computeTF(num_of_word2,stemmed_words2)
TF_Doc2

{'subject': 0.08333333333333333,
 'plan': 0.0,
 'fastest': 0.0,
 'across': 0.0,
 'acceler': 0.0,
 'machin': 0.08333333333333333,
 'math': 0.0,
 'action': 0.08333333333333333,
 'statist': 0.0,
 'learn': 0.08333333333333333,
 'one': 0.0,
 'specif': 0.08333333333333333,
 'grow': 0.0,
 'advanc': 0.0,
 'data': 0.08333333333333333,
 'guid': 0.0,
 'subsequ': 0.0,
 'uncov': 0.08333333333333333,
 'combin': 0.0,
 'special': 0.0,
 'matter': 0.08333333333333333,
 'make': 0.0,
 'organ': 0.08333333333333333,
 'field': 0.0,
 'made': 0.0,
 'everi': 0.0,
 'hidden': 0.08333333333333333,
 'industri': 0.0,
 'volum': 0.0,
 'sourc': 0.0,
 'expertis': 0.08333333333333333,
 'use': 0.0,
 'insight': 0.08333333333333333,
 'program': 0.0,
 'analyt': 0.0,
 'strateg': 0.0,
 'decis': 0.0,
 'scienc': 0.0}

In [110]:
# using the method that calculate the term frequency for document three
TF_Doc3 = computeTF(num_of_word3,stemmed_words3)
TF_Doc3

{'subject': 0.0,
 'plan': 0.0,
 'fastest': 0.0625,
 'across': 0.0625,
 'acceler': 0.0625,
 'machin': 0.0,
 'math': 0.0,
 'action': 0.0,
 'statist': 0.0,
 'learn': 0.0,
 'one': 0.0625,
 'specif': 0.0,
 'grow': 0.0625,
 'advanc': 0.0,
 'data': 0.1875,
 'guid': 0.0,
 'subsequ': 0.0625,
 'uncov': 0.0,
 'combin': 0.0,
 'special': 0.0,
 'matter': 0.0,
 'make': 0.0,
 'organ': 0.0,
 'field': 0.0625,
 'made': 0.0625,
 'everi': 0.0625,
 'hidden': 0.0,
 'industri': 0.0625,
 'volum': 0.0625,
 'sourc': 0.0625,
 'expertis': 0.0,
 'use': 0.0,
 'insight': 0.0,
 'program': 0.0,
 'analyt': 0.0,
 'strateg': 0.0,
 'decis': 0.0,
 'scienc': 0.0625}

In [111]:
# using the method that calculate the term frequency for the query 
TF_Q = computeTF(num_of_wordQ,stemmed_wordsQ)
TF_Q

{'data': 0.25, 'is': 0.25, 'science?': 0.25, 'what': 0.25}

In [112]:
# another way to calculate the term frequency 
from sklearn.feature_extraction.text import CountVectorizer

In [113]:
vectorizer = CountVectorizer()
tf_Doc1 = vectorizer.fit_transform(stemmed_words1)
feature_names_Doc1 = vectorizer.get_feature_names()
tf_dict_Doc1 = {}
for term, index in zip(feature_names_Doc1, range(len(feature_names_Doc1))):
    # Get the term frequency value for each document
    tf_values = tf_Doc1[:, index].toarray().flatten()
    tf_dict_Doc1[term] = tf_values
for term, tf_values in tf_dict_Doc1.items():
    print(f"{term}: {tf_values}")
#[data,scienc,combin,math,statist,special,program,advanc,analyt,insight,use,guid,decis,make,strateg,plan]

advanc: [0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0]
analyt: [0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0]
combin: [0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0]
data: [1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
decis: [0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0]
guid: [0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0]
insight: [0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0]
make: [0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0]
math: [0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0]
plan: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1]
program: [0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0]
scienc: [0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
special: [0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0]
statist: [0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0]
strateg: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0]
use: [0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0]


In [114]:
vectorizer = CountVectorizer()
tf_Doc2 = vectorizer.fit_transform(stemmed_words2)
feature_names_Doc2 = vectorizer.get_feature_names()
tf_dict_Doc2 = {}
for term, index in zip(feature_names_Doc2, range(len(feature_names_Doc2))):
    # Get the term frequency value for each document
    tf_values = tf_Doc2[:, index].toarray().flatten()
    tf_dict_Doc2[term] = tf_values
for term, tf_values in tf_dict_Doc2.items():
    print(f"{term}: {tf_values}")
#[machin,learn,specif,subject,matter,expertis,uncov,action,insight,hidden,organ,data]

action: [0 0 0 0 0 0 0 1 0 0 0 0]
data: [0 0 0 0 0 0 0 0 0 0 0 1]
expertis: [0 0 0 0 0 1 0 0 0 0 0 0]
hidden: [0 0 0 0 0 0 0 0 0 1 0 0]
insight: [0 0 0 0 0 0 0 0 1 0 0 0]
learn: [0 1 0 0 0 0 0 0 0 0 0 0]
machin: [1 0 0 0 0 0 0 0 0 0 0 0]
matter: [0 0 0 0 1 0 0 0 0 0 0 0]
organ: [0 0 0 0 0 0 0 0 0 0 1 0]
specif: [0 0 1 0 0 0 0 0 0 0 0 0]
subject: [0 0 0 1 0 0 0 0 0 0 0 0]
uncov: [0 0 0 0 0 0 1 0 0 0 0 0]


In [115]:
vectorizer = CountVectorizer()
tf_Doc3 = vectorizer.fit_transform(stemmed_words3)
feature_names_Doc3 = vectorizer.get_feature_names()
tf_dict_Doc3 = {}
for term, index in zip(feature_names_Doc3, range(len(feature_names_Doc3))):
    # Get the term frequency value for each document
    tf_values = tf_Doc3[:, index].toarray().flatten()
    tf_dict_Doc3[term] = tf_values
for term, tf_values in tf_dict_Doc3.items():
    print(f"{term}: {tf_values}")
#[acceler,volum,data,sourc,subsequ,data,made,data,scienc,one,fastest,grow,field,across,everi,industri]

acceler: [1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
across: [0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0]
data: [0 0 1 0 0 1 0 1 0 0 0 0 0 0 0 0]
everi: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0]
fastest: [0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0]
field: [0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0]
grow: [0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0]
industri: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1]
made: [0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0]
one: [0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0]
scienc: [0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0]
sourc: [0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0]
subsequ: [0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0]
volum: [0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0]


In [116]:
vectorizer = CountVectorizer()
tf_Q = vectorizer.fit_transform(stemmed_wordsQ)
feature_names_Q = vectorizer.get_feature_names()
tf_dict_Q = {}
for term, index in zip(feature_names_Q, range(len(feature_names_Q))):
    # Get the term frequency value for each document
    tf_values = tf_Q[:, index].toarray().flatten()
    tf_dict_Q[term] = tf_values
for term, tf_values in tf_dict_Q.items():
    print(f"{term}: {tf_values}")
#[what,is,data,science]

data: [0 0 1 0]
is: [0 1 0 0]
science: [0 0 0 1]
what: [1 0 0 0]


# Inverse Document Frequency (IDF)
Inverse Document Frequency (IDF) is a weight indicating how commonly a word is used. The more frequent its usage across documents, the lower its score. The lower the score, the less important the word becomes.

In [117]:
#IDF, this method calculate the inverse document frequency  
def computeIDF(Documents):
  import math
  N = len(Documents)
  IDF_Dict = dict.fromkeys(Documents[0].keys(),0)
  for Document in Documents:
    for word,val in Document.items():
      if val>0:
        IDF_Dict[word]+=1
  for word,val in IDF_Dict.items():
    IDF_Dict[word] = math.log (N /float(val))
  return IDF_Dict

In [118]:
# using the method to calculate the inverse document frequency for the documents 
IDFs = computeIDF([num_of_word1,num_of_word2,num_of_word3])
IDFs

{'subject': 1.0986122886681098,
 'plan': 1.0986122886681098,
 'fastest': 1.0986122886681098,
 'across': 1.0986122886681098,
 'acceler': 1.0986122886681098,
 'machin': 1.0986122886681098,
 'math': 1.0986122886681098,
 'action': 1.0986122886681098,
 'statist': 1.0986122886681098,
 'learn': 1.0986122886681098,
 'one': 1.0986122886681098,
 'specif': 1.0986122886681098,
 'grow': 1.0986122886681098,
 'advanc': 1.0986122886681098,
 'data': 0.0,
 'guid': 1.0986122886681098,
 'subsequ': 1.0986122886681098,
 'uncov': 1.0986122886681098,
 'combin': 1.0986122886681098,
 'special': 1.0986122886681098,
 'matter': 1.0986122886681098,
 'make': 1.0986122886681098,
 'organ': 1.0986122886681098,
 'field': 1.0986122886681098,
 'made': 1.0986122886681098,
 'everi': 1.0986122886681098,
 'hidden': 1.0986122886681098,
 'industri': 1.0986122886681098,
 'volum': 1.0986122886681098,
 'sourc': 1.0986122886681098,
 'expertis': 1.0986122886681098,
 'use': 1.0986122886681098,
 'insight': 0.4054651081081644,
 'progra

# TF-IDF models
TF-IDF stands for term frequency-inverse document frequency and it is a measure; TF-IDF(t,d,D) = tf(t, d) . idf (t,D) 

In [119]:
from sklearn.feature_extraction.text import TfidfVectorizer#to calculate the tf-idf 

In [128]:
#calculate TF-IDF for document one 
Doc1_stem = [' '.join(stemmed_words1)]
vectorizer_Doc1 = TfidfVectorizer()
Doc1_TFidf = vectorizer_Doc1.fit_transform(Doc1_stem)
print(Doc1_TFidf.toarray())
#data,scienc,combin,math,statist,special,program,advanc,analyt,insight,use,guid,decis,make,strateg,plan]

[[0.25 0.25 0.25 0.25 0.25 0.25 0.25 0.25 0.25 0.25 0.25 0.25 0.25 0.25
  0.25 0.25]]


In [129]:
#calculate TF-IDF for document two
Doc2_stem = [' '.join(stemmed_words2)]
vectorizer_Doc2 = TfidfVectorizer()
Doc2_TFidf = vectorizer_Doc2.fit_transform(Doc2_stem)
print(Doc2_TFidf.toarray())
#[machin,learn,specif,subject,matter,expertis,uncov,action,insight,hidden,organ,data]

[[0.28867513 0.28867513 0.28867513 0.28867513 0.28867513 0.28867513
  0.28867513 0.28867513 0.28867513 0.28867513 0.28867513 0.28867513]]


In [130]:
#calculate TF-IDF for document three
Doc3_stem = [' '.join(stemmed_words3)]
vectorizer_Doc3 = TfidfVectorizer()
Doc3_TFidf = vectorizer_Doc3.fit_transform(Doc3_stem)
print(Doc3_TFidf.toarray())
#[acceler,volum,data,sourc,subsequ,made,scienc,one,fastest,grow,field,across,everi,industri]

[[0.21320072 0.21320072 0.63960215 0.21320072 0.21320072 0.21320072
  0.21320072 0.21320072 0.21320072 0.21320072 0.21320072 0.21320072
  0.21320072 0.21320072]]


In [131]:
#calculate TF-IDF for the query 
Q_stem = [' '.join(stemmed_wordsQ)]
vectorizer_Q = TfidfVectorizer()
Q_TFidf = vectorizer_Q.fit_transform(Q_stem)
print(Q_TFidf.toarray())
#[what,is,data,science]

[[0.5 0.5 0.5 0.5]]


# Calculate the cosine similarity:
Influence and relevancy of documents with user query is measured using cosine similarity under a vector space where set of documents is considered as a set of vectors.

In [124]:
#to calculate the cosine similarity we used a library that exist so it can calculate it and rank the results 
from sklearn.metrics.pairwise import cosine_similarity
prepro_query = ' '.join(stemmed_wordsQ)

In [125]:
#calculating the cosine similarity between document one and the query 
prepro_Doc1 = ' '.join(stemmed_words1)

document1 = [prepro_Doc1, prepro_query]
# Initialize TfidfVectorizer
vectorizer = TfidfVectorizer()

# Fit and transform the vectorizer on the documents
tfidf_matrix= vectorizer.fit_transform(document1)

# Calculate cosine similarity between Doc1 and the query
cosine_similarities = cosine_similarity(tfidf_matrix[0], tfidf_matrix[1])

# Print the cosine similarity score
print(f"Cosine Similarity between Doc1 and the query: {cosine_similarities[0][0]}")


Cosine Similarity between Doc1 and the query: 0.06865710564998491


In [126]:
#calculating the cosine similarity between document two and the query 
prepro_Doc2 = ' '.join(stemmed_words2)

document2 = [prepro_Doc2, prepro_query]
# Initialize TfidfVectorizer
vectorizer = TfidfVectorizer()

# Fit and transform the vectorizer on the documents
tfidf_matrix = vectorizer.fit_transform(document2)

# Calculate cosine similarity between Doc1 and the query
cosine_similarities = cosine_similarity(tfidf_matrix[0], tfidf_matrix[1])

# Print the cosine similarity score
print(f"Cosine Similarity between Doc2 and the query: {cosine_similarities[0][0]}")

Cosine Similarity between Doc2 and the query: 0.07970251181266945


In [127]:
#calculating the cosine similarity between document three and the query 
prepro_Doc3 = ' '.join(stemmed_words3)

document3 = [prepro_Doc3, prepro_query]
# Initialize TfidfVectorizer
vectorizer = TfidfVectorizer()

# Fit and transform the vectorizer on the documents
tfidf_matrix = vectorizer.fit_transform(document3)

# Calculate cosine similarity between Doc1 and the query
cosine_similarities = cosine_similarity(tfidf_matrix[0], tfidf_matrix[1])

# Print the cosine similarity score
print(f"Cosine Similarity between Doc3 and the query: {cosine_similarities[0][0]}")

Cosine Similarity between Doc3 and the query: 0.19357302867354925


# Conclusion (the rank of documents)
In conclusion: after comparing each document with the query the result we got from calculating the cosine similarity and then ranking them in a descending order, so as the results show after ranking the documents, document three got the highest score.

Ranked Documents based on Cosine Similarity:

Document 3: Cosine Similarity = 0.19357302867354925

Document 2: Cosine Similarity = 0.07970251181266945

Document 1: Cosine Similarity = 0.06865710564998491